// unit/dist/does.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Does {
    /**
     * Assert that input is a string starting with the specified string.
     */
    static startWith(str) {
        return (input) => {
            if (typeof input === "string" && input.startsWith(str)) {
                return null;
            }
            return `Expected '${input}' to start with '${str}'`;
        };
    }
    /**
     * Assert that input is a string ending with the specified string.
     */
    static endWith(str) {
        return (input) => {
            if (typeof input === "string" && input.endsWith(str)) {
                return null;
            }
            return `Expected '${input}' to end with '${str}'`;
        };
    }
    /**
     * Assert that input matches specified patterns.
     * @param {RegExp[]} patterns The patterns to match.
     */
    static match(...patterns) {
        return (input) => {
            if (typeof input === "string") {
                let pass = true;
                for (const pattern of patterns) {
                    if (!pattern.test(input)) {
                        pass = false;
                        break;
                    }
                }
                if (pass) {
                    return null;
                }
            }
            return `Expected '${input}' to match patterns '${patterns}'`;
        };
    }
    /**
     * Assert that input is an array containing the specified item.
     * @param {*} item The item to verify the existance of.
     */
    static include(item) {
        return (input) => {
            if (Array.isArray(input) && input.includes(item)) {
                return null;
            }
            return `Expected '${input}' to include '${item}'`;
        };
    }
    /**
     * Assert that input entity contains a certain property.
     * @param {string} name The name of the property to verify the existance of.
     */
    static haveProperty(name) {
        return (input) => {
            if (name in input) {
                return null;
            }
            return `Expected '${input}' to have property '${name}'`;
        };
    }
    /**
     * Assert that input has the specified length property value.
     * @param {number} length The length to assert.
     */
    static haveLength(length) {
        return (input) => {
            if (input.length === length) {
                return null;
            }
            return `Expected '${input}' to have a length of '${length}' but got '${input.length}'`;
        };
    }
    /**
     * Assert that input has the specified size property value.
     * @param {number} length The size to assert.
     */
    static haveSize(size) {
        return (input) => {
            if (input.size === size) {
                return null;
            }
            return `Expected '${input}' to have a size of '${size}' but got '${input.size}'`;
        };
    }
}
exports.default = Does;
