// unit/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const is_1 = require("./is");
exports.Is = is_1.Is;
const assert_1 = __importDefault(require("./assert"));
exports.Assert = assert_1.default;
const runner_1 = __importStar(require("./runner"));
exports.Runner = runner_1.default;
exports.JsType = runner_1.JsType;
const decorators_1 = require("./decorators");
exports.unit = decorators_1.unit;
exports.test = decorators_1.test;
exports.feed = decorators_1.feed;
exports.target = decorators_1.target;
const does_1 = __importDefault(require("./does"));
exports.Does = does_1.default;
const mock_1 = __importDefault(require("./mock"));
exports.Mock = mock_1.default;
const stopwatch_1 = __importDefault(require("./stopwatch"));
exports.Stopwatch = stopwatch_1.default;
const dataSet_1 = __importStar(require("./dataSet"));
exports.DataSet = dataSet_1.default;
exports.OmitConflictingType = dataSet_1.OmitConflictingType;
